pkg/restore/restore.go

```diff
@@ -2195,3 +2208,40 @@ func (ctx *restoreContext) processUpdateResourcePolicy(fromCluster, fromClusterW
 	}
 	return warnings, errs
 }
+
+// function to process existingResourcePolicy as recreate, tries to delete object first before restoring obj
+func (ctx *restoreContext) processRecreateResourcePolicy(fromCluster, fromClusterWithLabels, obj *unstructured.Unstructured, namespace string, resourceClient client.Dynamic) (warnings, errs results.Result) {
+	ctx.log.Infof("restore API has existingResourcePolicy defined as recreate , executing restore workflow accordingly for changed resource %s %s ", obj.GroupVersionKind().Kind, kube.NamespaceAndName(fromCluster))
+	ctx.log.Infof("attempting recreate on %s %q", fromCluster.GetKind(), fromCluster.GetName())
+	// try to delete the object in cluster
+	err := resourceClient.Delete(obj.GetName(), metav1.DeleteOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			ctx.log.Errorf("delete attempt failed for %s %s: %v", fromCluster.GroupVersionKind(), kube.NamespaceAndName(fromCluster), err)
+		}
+	}
+	// wait up to 2 minutes until object does not exists in cluster
+	for timeStarted := time.Now(); apierrors.IsNotFound(err) || time.Now().After(timeStarted.Add(2*time.Minute)); {
+		_, err = resourceClient.Get(obj.GetName(), metav1.GetOptions{})
+		if !apierrors.IsNotFound(err) {
+			ctx.log.Warnf("get attempt to check object is gone failed for %s %s: %v", fromCluster.GroupVersionKind(), kube.NamespaceAndName(fromCluster), err)
+			time.Sleep(10 * time.Second)
+			continue
+		}
+		break
```

Comment:
how often are finalizer workloads going to restore/delete cleanly with just a simple finalizer removal?

I think most controllers with finalizer just need extra time to process deletion and go away on its own.. others are intentional deletion blocks that need more attention for deletion.

Reply:
hmm, I get your point. I think we can skip it then as part of the current work.
My thought around finalizers stemmed from the points I have put in the comment above.

resolving this for now.

Reply:
Similarly, I think there are more cases that the deletion of objects never succeeds in the recreate situation because of various unpredicted mechanisms, i.e., orders of deletion. Here is one example:

- As the current restore order, the restore of VolumeSnapshot objects is prior to PVC objects
- Suppose there is a PVC with its DataSource referring to a VolumeSnapshot object
- If we delete the VolumeSnapshot object first, it will always fail and complaining that this operation is blocked because a PVC is referring it.
- As far as I remember, the VS object is not marked as on-deletion in this case, that is, even though we delete the PVC later, the VS will not be deleted

Reply:
@kaovilai "others are intentional deletion blocks that need more attention for deletion." -- this could be a problem for restore. Following my comment above about implementing this after we have per-resource-type overrides, maybe we could have 2 different policies:
1) recreate-if-no-finalizers -- this would attempt to patch, and if patch failed, would delete-and-recreate only if there are no finalizers on the resource.
2) recreate-and-delete-finalizers -- this would attempt to patch and if patch failed, would remove finalizers before attempting to delete

The latter is potentially dangerous, so we would not want to allow it as an overall policy, but might make sense for some specific named resource/crd that the user knows is safe to delete-and-recreate in this way.